include/deal.II/base/tensor.h

```diff
@@ -1487,348 +1479,221 @@ operator * (const Tensor<rank_1, dim, Number> &src1,
- * @f]
- *
- * @relates Tensor
- * @author Matthias Maier, 2015
- */
-template <int rank_1, int rank_2, int dim,
-          typename T1, typename T2, typename T3>
-typename ProductType<T1, typename ProductType<T2, T3>::type>::type
-contract3 (const Tensor<rank_1, dim, T1> &left,
-           const Tensor<rank_1 + rank_2, dim, T2> &middle,
-           const Tensor<rank_2, dim, T3> &right)
-{
-  typedef typename ProductType<T1, typename ProductType<T2, T3>::type>::type
-  return_type;
-  return TensorAccessors::contract3<rank_1, rank_2, dim, return_type>(
-           left, middle, right);
-}
-
-
-/**
- * The outer product of two tensors of @p rank_1 and @p rank_2: Returns a
- * tensor of rank $(\text{rank_1} + \text{rank_2})$:
+ * Generic contraction of a pair of indices of two tensors of arbitrary
+ * rank: Return a tensor of rank $(\text{rank_1} + \text{rank_2} - 2)$ that
+ * is the contraction of index @p index_1 of a tensor @p src1 of rank
+ * @p rank_1 with the index @p index_2 of a tensor @p src2 of rank @p rank_2:
  * @f[
  *   \text{result}_{i_1,..,i_{r1},j_1,..,j_{r2}}
- *   = \text{left}_{i_1,..,i_{r1}}\,\text{right}_{j_1,..,j_{r2}.}
+ *   = \sum_{k}
+ *     \text{left}_{i_1,..,k,..,i_{r1}}
+ *     \text{right}_{j_1,..,k,..,j_{r2}}
  * @f]
  *
+ * If for example the first index (<code>index_1==0</code>) of a tensor
+ * <code>t1</code> shall be contracted with the third index
+ * (<code>index_2==2</code>) of a tensor <code>t2</code>, the invocation of
+ * this function is
+ * @code
+ *   contract<0, 2>(t1, t2);
+ * @endcode
+ *
+ * @note The position of the index is counted from 0, i.e.,
+ * $0\le\text{index_i}<\text{range_i}$.
+ *
+ * @note In case the contraction yields tensor of rank 0 the scalar
+ * number is returned as an unwrapped number type.
+ *
  * @relates Tensor
+ * @author Matthias Maier, 2015
  */
-template <int rank_1, int rank_2, int dim,
+template <int index_1, int index_2,
+          int rank_1, int rank_2, int dim,
           typename Number, typename OtherNumber>
 inline
-Tensor<rank_1 + rank_2, dim, typename ProductType<Number, OtherNumber>::type>
-outer_product(const Tensor<rank_1, dim, Number> &src1,
-              const Tensor<rank_2, dim, OtherNumber> &src2)
-{
-  typename Tensor<rank_1 + rank_2, dim, typename ProductType<Number, OtherNumber>::type>::tensor_type result;
-  TensorAccessors::contract<0, rank_1, rank_2, dim>(result, src1, src2);
-  return result;
-}
-
-
-//@}
-/**
- * @name To be refactored
- */
-//@{
+typename Tensor<rank_1 + rank_2 - 2, dim, typename ProductType<Number, OtherNumber>::type>::tensor_type
+contract (const Tensor<rank_1, dim, Number> &src1,
+          const Tensor<rank_2, dim, OtherNumber> &src2)
+{
+#ifdef DEAL_II_WITH_CXX11
+  static_assert(0 <= index_1 && index_1 < rank_1,
+                "The specified index_1 must lie within the range [0,rank_1)");
+  static_assert(0 <= index_2 && index_2 < rank_2,
+                "The specified index_2 must lie within the range [0,rank_2)");
```

Comment:
I think this would be worthwhile having in non-C++11 mode as well. Can you either make it a regular `Assert` or put the regular `Assert` in an `#else` branch?